pkg/virt-controller/watch/snapshot.go

```diff
@@ -0,0 +1,691 @@
+						"VolumeSnapshot %s no longer exists",
+						*volmeBackup.VolumeSnapshotName,
+					)
+					ready = false
+					deletedSnapshots = append(deletedSnapshots, *volmeBackup.VolumeSnapshotName)
+					continue
+				}
+
+				log.Log.Infof("Attempting to create VolumeSnapshot %s", *volmeBackup.VolumeSnapshotName)
+
+				sc := volmeBackup.PersistentVolumeClaim.Spec.StorageClassName
+				if sc == nil {
+					return fmt.Errorf("%s/%s VolumeSnapshot requested but no storage class",
+						content.Namespace, volmeBackup.PersistentVolumeClaim.Name)
+				}
+
+				volumeSnapshotClass, err := ctrl.getVolumeSnapshotClass(*sc)
+				if err != nil {
+					log.Log.Warningf("Couldn't find VolumeSnapshotClass for %s", *sc)
+					return err
+				}
+
+				t := true
+				snapshot := &k8ssnapshotv1beta1.VolumeSnapshot{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: *volmeBackup.VolumeSnapshotName,
+						OwnerReferences: []metav1.OwnerReference{
+							{
+								APIVersion:         vmsnapshotv1alpha1.SchemeGroupVersion.String(),
+								Kind:               "VirtualMachineSnapshotContent",
+								Name:               content.Name,
+								UID:                content.UID,
+								Controller:         &t,
+								BlockOwnerDeletion: &t,
+							},
+						},
+					},
+					Spec: k8ssnapshotv1beta1.VolumeSnapshotSpec{
+						Source: k8ssnapshotv1beta1.VolumeSnapshotSource{
+							PersistentVolumeClaimName: &volmeBackup.PersistentVolumeClaim.Name,
+						},
+						VolumeSnapshotClassName: &volumeSnapshotClass,
+					},
+				}
+
+				_, err = ctrl.client.KubernetesSnapshotClient().SnapshotV1beta1().
+					VolumeSnapshots(content.Namespace).
+					Create(snapshot)
+				if err != nil {
+					if !errors.IsAlreadyExists(err) {
+						return err
+					}
+				} else {
+					ctrl.recorder.Eventf(
+						snapshot,
+						corev1.EventTypeNormal,
+						volumeSnapshotCreateEvent,
+						"Successfully created VolumeSnapshot %s",
+						snapshot.Name,
+					)
+				}
+
+				ready = false
+				continue
+			} else if err != nil {
+				return err
+			}
+
+			vs := obj.(*k8ssnapshotv1beta1.VolumeSnapshot)
+
+			if vs.Status != nil {
+				if vs.Status.ReadyToUse == nil || *vs.Status.ReadyToUse == false {
+					ready = false
+				}
+
+				if vs.Status.Error != nil {
+					ready = false
+					vmSnapshotError = &vmsnapshotv1alpha1.VirtualMachineSnapshotError{
+						Time:    vs.Status.Error.Time,
+						Message: vs.Status.Error.Message,
+					}
+
+					break
+				}
+			}
+		}
+	}
+
+	if vmSnapshotError == nil && len(deletedSnapshots) > 0 {
+		message := fmt.Sprintf("VolumeSnapshots (%s) missing", strings.Join(deletedSnapshots, ","))
+		vmSnapshotError = &vmsnapshotv1alpha1.VirtualMachineSnapshotError{
+			Time:    currentTime(),
+			Message: &message,
+		}
+	}
+
+	if !ready {
+		creationTime = nil
+	}
+
+	contentCpy := content.DeepCopy()
+	contentCpy.Status.ReadyToUse = &ready
+	contentCpy.Status.CreationTime = creationTime
+	contentCpy.Status.Error = vmSnapshotError
```

Comment:
here's a thought, since it's possible multiple errors could exist, it might be helpful to use conditions here which would let you express all the error conditions that might be occurring.

Reply:
Yeah I've been going back and forth on this